src/ProductConstructionService/ProductConstructionService.Api/Metrics/MetricRecorder.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using ProductConstructionService.Api.Queue.Jobs;
+using ProductConstructionService.ServiceDefaults;
+
+namespace ProductConstructionService.Api.Metrics;
+
+public class MetricRecorder(IMeterFactory meterFactory, ILogger<MetricRecorder> logger) : IMetricRecorder
+{
+    private readonly ILogger<MetricRecorder> _logger = logger;
+    private readonly Meter _meter = meterFactory.Create(MetricConsts.JobMeterName);
+    private readonly Dictionary<string, JobMetricRecorderInstruments> _jobMetricRecorderInstruments = new();
+
+    public IMetricRecorderScope RecordJob(Job job)
+    {
+        var instrumentName = $"job.{job.Type}";
+
+        if (!_jobMetricRecorderInstruments.TryGetValue(instrumentName, out var instruments))
+        {
+            var histogramName = $"job.{job.Type}.duration";
+            var successCounterName = $"job.{job.Type}.success.count";
+            var failureCounterName = $"job.{job.Type}.failure.count";
+
+            instruments = new(
+                _meter.CreateHistogram<long>(histogramName),
+                _meter.CreateCounter<int>(successCounterName),
+                _meter.CreateCounter<int>(failureCounterName));
+            _jobMetricRecorderInstruments.Add(instrumentName, instruments);
+        }
+
+        return new MetricRecorderScope(job, instruments, _logger);
+    }
+
+    private class MetricRecorderScope(Job job, JobMetricRecorderInstruments instruments, ILogger logger) : IMetricRecorderScope
```

Comment:
This class doesn't have `Job` in its name but is tied to a `Job`. I was thinking this would be more generic:
- `private class MetricScope`
- Not dependent on `job` but just have `_isSuccessful` because it can be any operation (like git clone)
- `JobMetricRecorderInstruments` could also be just generic or it would receive a finalizer Action again.

It think for most metrics, we will only measure success and duration so this all could be abstract, only the name of the metric differs really and that can be passed from `RecordJob`.

Because if this would not have the logging statement, it would be generic enough. We could log elsewhere (int JobScope for instance or JobProcessor etc).